src/java/org/jivesoftware/openfire/plugin/rest/controller/SystemController.java

```diff
@@ -0,0 +1,273 @@
+     */
+    public boolean isStarted() {
+        return XMPPServer.getInstance().isStarted();
+    }
+
+    /**
+     * Determine if clustering has been fully started, if clustering is enabled.
+     *
+     * When clustering is not enabled, this method will always return 'true'.
+     *
+     * @return true if Openfire has started, otherwise false.
+     */
+    public boolean hasClusteringStartedWhenEnabled() {
+        if (!ClusterManager.isClusteringEnabled()) {
+            return true;
+        }
+        return ClusterManager.isClusteringStarted();
+    }
+
+    /**
+     * Checks if the plugin manager has finished starting all plugins that were available at boot-time.
+     *
+     * Note that the return value does not indicate that plugins succeeded or failed to start: a 'true' return value
+     * indicates only that the plugin manager finished its initial attempt to start all plugins.
+     *
+     * @return True when all (initial) plugins have been started.
+     */
+    public boolean hasPluginManagerExecuted() {
+        return XMPPServer.getInstance().getPluginManager().isExecuted();
+    }
+
+    /**
+     * Verifies that a connection listener for the provided type and encryption level is disabled, or ready to accept connections.
+     *
+     * @param connectionType The type of connection for which to check state
+     * @param encrypted true when the direct-TLS encrypted variant of the listeners is to be checked, otherwise false.
+     * @return True when the connection listener is ready to accept connections, or is disabled.
+     */
+    public boolean isConnectionListenerStartedWhenEnabled(@Nonnull final ConnectionType connectionType, final boolean encrypted) {
+        switch (connectionType) {
+            case BOSH_C2S:
+                if (encrypted) {
+                    return !HttpBindManager.HTTP_BIND_ENABLED.getValue() || HttpBindManager.HTTP_BIND_SECURE_PORT.getValue() <= 0 || HttpBindManager.getInstance().isHttpsBindActive();
+                } else {
+                    return !HttpBindManager.HTTP_BIND_ENABLED.getValue() || HttpBindManager.HTTP_BIND_PORT.getValue() <= 0 || HttpBindManager.getInstance().isHttpBindActive();
+                }
+
+            case WEBADMIN:
+                // FIXME. See OF-2400
+                return true;
+
+            case SOCKET_S2S:
+                // FIXME. See OF-2400
+                return true;
+
+            default:
+                final ConnectionManagerImpl connectionManager = (ConnectionManagerImpl) XMPPServer.getInstance().getConnectionManager();
+                if (connectionManager == null) {
+                    return false;
+                }
+                final ConnectionListener listener = connectionManager.getListener(connectionType, encrypted);
+                if (listener == null) {
+                    return false;
+                }
+                return !listener.isEnabled() || listener.getSocketAcceptor() != null;
```

Comment:
Is this logic correct?
If the listener is _not_ enabled OR if the listener has a MINA NioSocketAcceptor, then return True?
Feels like is should be: if it _is_ enabled (which defaults to true if it isn't configurable) AND has a MINA NioSocketAcceptor, then return True

Reply:
I did this on purpose. Putting it differently, when _disabled_, the listener immediately is in the state that we expect it to be in. My reasoning was as follows: if the listener is disabled, the check to see if it is 'ready' should pass. If the check does not pass when a listener is disabled, then this test will always indicate that the server isn't ready to be used.

I tried to give the method a name that reflects this, but I won't argue that it's confusing. Assuming that you agree with my logic, do you see a way to make things more logical?

Reply:
That does make more sense.